File: app/src/main/java/de/datlag/hotdrop/MainActivity.java
package de.datlag.hotdrop;

import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatEditText;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.LinearLayoutCompat;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.content.ContextCompat;
import androidx.core.view.ViewCompat;

import com.adroitandroid.near.model.Host;
import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.MobileAds;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.leinardi.android.speeddial.SpeedDialActionItem;
import com.leinardi.android.speeddial.SpeedDialView;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

import de.datlag.hotdrop.extend.AdvancedActivity;
import de.datlag.hotdrop.firebase.StorageManager;
import de.datlag.hotdrop.fragment.ChooseDeviceFragment;
import de.datlag.hotdrop.fragment.SearchDeviceFragment;
import de.datlag.hotdrop.manager.InfoPageManager;
import de.datlag.hotdrop.manager.PermissionManager;
import de.datlag.hotdrop.manager.SettingsManager;
import de.datlag.hotdrop.p2p.DiscoverHost;
import de.interaapps.firebasemanager.auth.AnonymousAuth;
import de.interaapps.firebasemanager.auth.EmailAuth;
import de.interaapps.firebasemanager.auth.GoogleAuth;
import de.interaapps.firebasemanager.auth.OAuth;
import de.interaapps.firebasemanager.auth.OAuthEnum;
import de.interaapps.firebasemanager.core.FirebaseManager;
import io.noties.markwon.Markwon;
import io.noties.markwon.ext.strikethrough.StrikethroughPlugin;
import io.noties.markwon.html.HtmlPlugin;
import lombok.Getter;

public class MainActivity extends AdvancedActivity implements ChooseDeviceFragment.OnFragmentInteractionListener {

    private AdvancedActivity activity;
    private CoordinatorLayout coordinatorLayout;
    private Toolbar toolbar;
    private SpeedDialView speedDialView;
    private AppCompatImageView backgroundImage;
    private AppCompatImageView tuneButton;
    private AppCompatImageView revealButton;
    private AppCompatImageView reverseRevealButton;
    private FrameLayout infoLayout;
    private FrameLayout mainLayout;
    private AppCompatImageView githubIcon;
    private AppCompatImageView codeIcon;
    private AppCompatImageView helpIcon;
    private PermissionManager permissionManager;
    private InfoPageManager infoPageManager;
    private SettingsManager settingsManager;
    private FirebaseManager firebaseManager;
    private StorageManager storageManager;
    private AdView adView;
    private AdRequest adRequest;
    private GoogleAuth googleAuth;
    private EmailAuth emailAuth;
    private OAuth githubAuth;
    private AnonymousAuth anonymousAuth;
    private Markwon markwon;
    @Getter
    private SearchDeviceFragment searchDeviceFragment;
    private static final int downloadID = ViewCompat.generateViewId();
    private static final int uploadID = ViewCompat.generateViewId();
    private ArrayList<SpeedDialActionItem> menuItems = new ArrayList<>();
    private boolean adLoaded = false;
    private boolean speedDialMoved = false;
    private DiscoverHost discoverHost;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        activity = MainActivity.this;

        initImportant();
        initViews();
        initViewLogic();
        initLogic();
        initListener();
        initDeeplink();
        initReceive();
    }

    private void initImportant() {
        createShortcuts();
        permissionManager = new PermissionManager(activity);
        permissionManager.check();

        MobileAds.initialize(activity, getString(R.string.admob_app_id));
        adRequest = new AdRequest.Builder().build();

        firebaseManager = new FirebaseManager(activity);
        googleAuth = new GoogleAuth(activity, getString(R.string.default_web_client_id));
        emailAuth = new EmailAuth(activity);
        githubAuth = new OAuth(activity, OAuthEnum.GITHUB);
        anonymousAuth = new AnonymousAuth(activity);
        firebaseManager.addLogin(googleAuth);
        firebaseManager.addLogin(emailAuth);
        firebaseManager.addLogin(githubAuth);
        firebaseManager.addLogin(anonymousAuth);
        settingsManager = new SettingsManager(activity, firebaseManager);
        storageManager = new StorageManager(activity, firebaseManager, settingsManager);

        discoverHost = new DiscoverHost(activity);
    }

    private void initViews() {
        coordinatorLayout = findViewById(R.id.coordinator);
        toolbar = findViewById(R.id.toolbar);
        speedDialView = findViewById(R.id.speedDial);
        backgroundImage = findViewById(R.id.background_image);
        tuneButton = findViewById(R.id.tune_button);
        revealButton = findViewById(R.id.reveal_button);
        infoLayout = findViewById(R.id.info_layout);
        mainLayout = findViewById(R.id.frame);
        reverseRevealButton = findViewById(R.id.reverse_reveal);
        githubIcon = findViewById(R.id.github_icon);
        codeIcon = findViewById(R.id.code_icon);
        helpIcon = findViewById(R.id.help_icon);
        adView = findViewById(R.id.adView);
    }

    private void initViewLogic() {
        Glide.with(activity)
                .load(ContextCompat.getDrawable(activity, R.drawable.circles))
                .centerCrop()
                .apply(new RequestOptions().fitCenter())
                .override(getResources().getDisplayMetrics().widthPixels, getResources().getDisplayMetrics().heightPixels)
                .into(backgroundImage);

        menuItems.add(new SpeedDialActionItem.Builder(downloadID, R.drawable.ic_cloud_download_white_24dp)
                .setFabImageTintColor(Color.WHITE)
                .create());
        menuItems.add(new SpeedDialActionItem.Builder(uploadID, R.drawable.ic_cloud_upload_white_24dp)
                .setFabImageTintColor(Color.WHITE)
                .create());
        speedDialView.addAllActionItems(menuItems);

        adView.loadAd(adRequest);
    }

    private void initLogic() {
        setSupportActionBar(toolbar);
        checkConfiguration();
        markwon = Markwon.builder(activity)
                .usePlugin(StrikethroughPlugin.create())
                .usePlugin(HtmlPlugin.create())
                .build();

        searchDeviceFragment = SearchDeviceFragment.newInstance(activity, discoverHost);
        switchFragment(searchDeviceFragment, R.id.fragment_view);

        infoPageManager = new InfoPageManager(activity, mainLayout, infoLayout);
        infoPageManager.setCodeIcon(codeIcon);
        infoPageManager.setGithubIcon(githubIcon);
        infoPageManager.setHelpIcon(helpIcon);
        infoPageManager.init();
    }

    private void initListener() {
        adView.setAdListener(new AdListener(){
            @Override
            public void onAdLoaded() {
                super.onAdLoaded();
                adLoaded = true;
                moveSpeedDial(true);
            }

            @Override
            public void onAdClosed() {
                super.onAdClosed();
                adLoaded = false;
                moveSpeedDial(false);
            }

            @Override
            public void onAdFailedToLoad(int i) {
                super.onAdFailedToLoad(i);
                adLoaded = false;
                moveSpeedDial(false);
            }
        });

        revealButton.setOnClickListener((View v) -> {
                infoPageManager.start(false);
        });

        reverseRevealButton.setOnClickListener((View v) -> {
                infoPageManager.start(true);
        });

        speedDialView.setOnActionSelectedListener((SpeedDialActionItem actionItem) -> {
                if (actionItem.getId() == uploadID) {
                    storageManager.upload();
                } else if (actionItem.getId() == downloadID) {
                    LayoutInflater inflater = activity.getLayoutInflater();
                    LinearLayoutCompat linearLayoutCompat = new LinearLayoutCompat(activity);
                    inflater.inflate(R.layout.download_dialog, linearLayoutCompat);
                    AppCompatEditText appCompatEditText = linearLayoutCompat.findViewById(R.id.download_edittext);
                    activity.applyDialogAnimation(new MaterialAlertDialogBuilder(activity)
                            .setTitle("Download file")
                            .setView(linearLayoutCompat)
                            .setPositiveButton(activity.getString(R.string.okay), (dialog, which) -> {
                                storageManager.download(String.valueOf(appCompatEditText.getText()));
                            })
                            .setNegativeButton(activity.getString(R.string.cancel), null)
                            .create()).show();
                }
                return false;
        });

        tuneButton.setOnClickListener((View v) -> {
                settingsManager.open();
        });

    }

    private void initDeeplink() {
        Uri data = activity.getIntent().getData();
        if (data != null && data.isHierarchical()) {
            String url = activity.getIntent().getDataString();
            if (isURLValid(url)) {
                askDownload(url);
            }
        }
    }

    private void initReceive() {
        Intent intent = activity.getIntent();
        String action = intent.getAction();
        String type = intent.getType();

        if (action.equals(Intent.ACTION_SEND) && type != null) {
            if (type.equals("text/plain")) {
                String url = intent.getStringExtra(Intent.EXTRA_TEXT);
                if (isURLValid(url)) {
                    Log.e("Download", url);
                    askDownload(url);
                }
            } else {
                //ToDo: upload or search device
            }
        }
    }

    private void askDownload(String url) {
        activity.applyDialogAnimation(new MaterialAlertDialogBuilder(activity)
                .setTitle("Cloud File")
                .setMessage("Do you want to download the file from this link:\n\n"+url)
                .setPositiveButton(R.string.okay, (dialog, which) -> {
                    storageManager.download(url);
                })
                .setNegativeButton(R.string.cancel, (dialog, which) -> {
                    dialog.cancel();
                })
                .setNeutralButton("Copy Link", (dialog, which) -> {
                    activity.copyText("HotDrop Download URL", url);
                })
                .create()).show();
    }

    @Contract(pure = true)
    private boolean isURLValid(@NotNull String url) {
        return url.contains("gs://hotdrop-420.appspot.com");
    }

    private void moveSpeedDial(boolean up) {
        if (up) {
            if (!speedDialMoved && adLoaded) {
                ViewGroup.MarginLayoutParams params = (ViewGroup.MarginLayoutParams) speedDialView.getLayoutParams();
                params.setMargins(params.getMarginStart(), params.topMargin, params.getMarginEnd(), params.bottomMargin + getActionBarHeight());
                speedDialView.setLayoutParams(params);
                speedDialMoved = true;
            }
        } else {
            if (speedDialMoved) {
                ViewGroup.MarginLayoutParams params = (ViewGroup.MarginLayoutParams) speedDialView.getLayoutParams();
                params.setMargins(params.getMarginStart(), params.topMargin, params.getMarginEnd(), params.bottomMargin - getActionBarHeight());
                speedDialView.setLayoutParams(params);
                speedDialMoved = false;
            }
        }
    }

    @Override
    public void onChooseFragmentInteraction(Host host) {
        discoverHost.send(host, DiscoverHost.MESSAGE_REQUEST_START_TRANSFER.getBytes());
    }

    @Override
    protected void onOrientationLandscape() {
        moveSpeedDial(false);
    }

    @Override
    protected void onOrientationPortrait() {
        moveSpeedDial(true);
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (firebaseManager != null) {
            firebaseManager.onStart();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        moveSpeedDial(false);
        discoverHost.stopDiscovery();
    }

    @Override
    protected void onResume() {
        super.onResume();
        moveSpeedDial(true);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        activity.finishAffinity();
        discoverHost.stopDiscovery();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        googleAuth.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }
}

File: app/src/main/java/de/datlag/hotdrop/MainActivity.kt
package de.datlag.hotdrop

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.widget.FrameLayout
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.Toolbar
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.MobileAds
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.play.core.appupdate.AppUpdateManagerFactory
import com.google.android.play.core.install.model.AppUpdateType
import com.google.android.play.core.install.model.UpdateAvailability
import de.datlag.hotdrop.extend.AdvancedActivity
import de.datlag.hotdrop.fragment.ChooseDeviceFragment
import de.datlag.hotdrop.fragment.ChooseDeviceFragment.OnFragmentInteractionListener
import de.datlag.hotdrop.fragment.SearchDeviceFragment
import de.datlag.hotdrop.manager.InfoPageManager
import de.datlag.hotdrop.manager.PermissionManager
import de.datlag.hotdrop.manager.SettingsManager
import io.noties.markwon.Markwon
import io.noties.markwon.ext.strikethrough.StrikethroughPlugin
import io.noties.markwon.html.HtmlPlugin
import de.datlag.hotdrop.fragment.SearchDeviceFragment.SearchCallback
import de.datlag.hotdrop.p2p.*
import de.datlag.hotdrop.util.FileUtil
import java.io.File

class MainActivity : AdvancedActivity(), OnFragmentInteractionListener, SearchCallback {
    private val activity: AdvancedActivity = this@MainActivity
    private val toolbar: Toolbar by bindView(R.id.toolbar)
    val backgroundImage: AppCompatImageView by bindView(R.id.background_image)
    private val tuneButton: AppCompatImageView by bindView(R.id.tune_button)
    private val revealButton: AppCompatImageView by bindView(R.id.reveal_button)
    private val reverseRevealButton: AppCompatImageView by bindView(R.id.reverse_reveal)
    private val infoLayout: FrameLayout by bindView(R.id.info_layout)
    private val mainLayout: FrameLayout by bindView(R.id.frame)
    private val githubIcon: AppCompatImageView by bindView(R.id.github_icon)
    private val codeIcon: AppCompatImageView by bindView(R.id.code_icon)
    private val helpIcon: AppCompatImageView by bindView(R.id.help_icon)
    private val adView: AdView by bindView(R.id.adView)


    private lateinit var permissionManager: PermissionManager
    private lateinit var infoPageManager: InfoPageManager
    private lateinit var settingsManager: SettingsManager
    private lateinit var adRequest: AdRequest
    private lateinit var markwon: Markwon

    lateinit var searchDeviceFragment: SearchDeviceFragment
    var chooseDeviceFragment: ChooseDeviceFragment? = null
    var showingChooseFragment = false
    private lateinit var hostDiscovery: HostDiscovery
    private lateinit var fileTransfer: FileTransfer


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initImportant()
        initViewLogic()
        initLogic()
        initListener()
        initReceive()
    }

    private fun initImportant() {
        createShortcuts()
        checkUpdate()

        permissionManager = PermissionManager(activity)
        permissionManager.check()
        MobileAds.initialize(activity, getString(R.string.admob_app_id))
        adRequest = AdRequest.Builder().build()
        hostDiscovery = HostDiscovery(activity)
    }

    private fun initViewLogic() {
        adView.loadAd(adRequest)
        settingsManager = SettingsManager(activity)
    }

    private fun initLogic() {
        setSupportActionBar(toolbar)
        checkConfiguration()
        markwon = Markwon.builder(activity)
                .usePlugin(StrikethroughPlugin.create())
                .usePlugin(HtmlPlugin.create())
                .build()
        searchDeviceFragment = SearchDeviceFragment.newInstance(activity)
        switchFragment(searchDeviceFragment, R.id.fragment_view)
        infoPageManager = InfoPageManager(activity, mainLayout, infoLayout)
        infoPageManager.codeIcon = codeIcon
        infoPageManager.githubIcon = githubIcon
        infoPageManager.helpIcon = helpIcon
        infoPageManager.init()

        fileTransfer = FileTransfer(activity)
    }

    private fun initListener() {
        revealButton.setOnClickListener { infoPageManager.start(false) }
        reverseRevealButton.setOnClickListener { infoPageManager.start(true) }

        tuneButton.setOnClickListener { settingsManager.open() }
    }

    private fun initReceive() {
        val intent = activity.intent
        val action = intent.action
        val type = intent.type
        if (action == Intent.ACTION_SEND && type != null) {
            TODO()
        }
    }

    private fun checkUpdate() {
        val appUpdateManager = AppUpdateManagerFactory.create(activity)
        val appUpdateInfoTask = appUpdateManager.appUpdateInfo

        appUpdateInfoTask.addOnSuccessListener {
            if(it.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE
                    && it.isUpdateTypeAllowed(AppUpdateType.IMMEDIATE)) {
                appUpdateManager.startUpdateFlowForResult(it, AppUpdateType.IMMEDIATE, activity, UPDATE_REQUEST_CODE)
            }
        }
    }

    override fun onChooseFragmentInteraction(host: Host) {
        /*
            activity.applyDialogAnimation(MaterialAlertDialogBuilder(activity)
                    .setTitle(host.name)
                    .setMessage("Request Connection with ${host.name}")
                    .setPositiveButton("Request") { _, _ ->
                        host.send("START_TRANSFER".toByteArray())
                    }
                    .setNeutralButton("Cancel", null)
                    .create()).show()
         */
        applyDialogAnimation(MaterialAlertDialogBuilder(activity)
                .setTitle(host.name)
                .setMessage("Do you want to send a File or a Folder to this device?\nFolders will be packed in a ZIP-Archiv and then sending the .zip file.")
                .setPositiveButton("File") { _, _ ->
                    FileUtil.chooseFile(activity, null, null, object: FileUtil.FileChooseCallback{
                        override fun onChosen(path: String?, file: File?) {
                            fileTransfer.send(host, file!!)
                        }
                    })
                }
                .setNegativeButton("Folder") { _, _ ->
                    FileUtil.chooseFolder(activity, null, object: FileUtil.FolderChooseCallback {
                        override fun onChosen(path: String?, file: File?) {
                            fileTransfer.send(host, FileUtil.folderToFile(activity, file!!))
                        }
                    })
                }
                .setNeutralButton(R.string.cancel, null)
                .create()).show()
    }

    override fun onStop() {
        super.onStop()
        hostDiscovery.stop()
    }

    override fun onDestroy() {
        super.onDestroy()
        hostDiscovery.stop()
        activity.finishAffinity()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if(requestCode == UPDATE_REQUEST_CODE) {
            if(resultCode != Activity.RESULT_OK) {
                activity.applyDialogAnimation(MaterialAlertDialogBuilder(activity)
                        .setTitle("Update")
                        .setMessage("Update flow failed!\nTry again or update in Google Play Store")
                        .setPositiveButton("Try again"){
                            _, _ -> checkUpdate()
                        }
                        .setNegativeButton("Google Play") {
                            _, _ -> browserIntent("https://play.google.com/store/apps/details?id=de.datlag.hotdrop")
                        }
                        .create()).show()
            }
        }
    }

    companion object {
        private const val UPDATE_REQUEST_CODE = 1337
    }

    override fun onSearchChanged(search: Boolean) {
        if (search) {
            hostDiscovery.start(object: DiscoveryCallback {
                override fun onHostsFound(hosts: Set<Host>) {
                    if (hosts.isEmpty()) {
                        switchSearchAndChoose(false)
                        return
                    }

                    if (chooseDeviceFragment == null) {
                        chooseDeviceFragment = ChooseDeviceFragment.newInstance(hosts)
                        switchFragment(chooseDeviceFragment!!, R.id.fragment_view)
                    } else {
                        chooseDeviceFragment!!.setHosts(hosts)
                    }
                    switchSearchAndChoose(true)
                }
            })
        } else {
            hostDiscovery.stop()
            switchSearchAndChoose(false)
        }
    }

    private fun switchSearchAndChoose(showChooseFragment: Boolean) {
        if (showChooseFragment && !showingChooseFragment) {
            chooseDeviceFragment?.let {
                switchFragment(it, R.id.fragment_view)
                showingChooseFragment = true
            }
        } else {
            switchFragment(searchDeviceFragment, R.id.fragment_view)
            showingChooseFragment = false
        }
    }
}